Improvements for Security doc (#7640)
mendix/docs: content/en/docs/refguide/modeling/app-explorer/security/_index.md

## 1 Introduction

> Security in Mendix has several aspects. One aspect is that your app is a project to work on in the backend, develop it and collaborate on it with your colleagues as a team. Yet you may want to control what each team member can change in the backend and this is managed in the Developer Portal. For more information, see [Building an App as a Team](#team) section below. 

Reviewer: I know what you are saying here, but I think it could be more direct. Is it OK if I make a suggestion to rewrite it?

Author: I tried, but I am not sure if I made it any better. Do you mean to rewrite the whole intro or just this paragraph?

Reviewer: I'll make some suggestions - but feel free to adapt or ignore them if you think they aren't an improvement.
```suggestion
Security in Mendix has several aspects which are all covered on this page.

Firstly, you may be working as a team to create the app. You will want to control who is on the team and what each member can do. This is managed in the Developer Portal. For more information, see [Building an App as a Team](#team) section below. 
```

If you want different end-users to see different parts of your running application and you want to prevent unauthorized access, these aspects are managed from Studio Pro. Access to forms, data and microflows can be limited to authorized users. For more information, see [Security in Studio Pro](#security-in-studio-pro) section below.

{{% alert color="info" %}}

> Security in Studio Pro does not include scanning files that end-users upload or download from your application for viruses and malware. For more information, see the [Scanning Uploaded Files for Malicious Content](/howto/security/best-practices-security/#scanning-for-malicious-content) section in *How to Implement Best Practices for App Security*. 

Reviewer: No - this is security in the Mendix app, not Studio Pro. It is talking about an end-user uploading a file (an image or some other data the company needs to process).
```suggestion
Security in Mendix apps does not include scanning files that end-users upload or download from your application for viruses and malware. For more information, see the [Scanning Uploaded Files for Malicious Content](/howto/security/best-practices-security/#scanning-for-malicious-content) section in *How to Implement Best Practices for App Security*. 
```

{{% /alert %}}

> You can also set app authentication and add SSO modules to your app using Mendix Marketplace. For more information, see [App Authentication](#authentication) section below. 

Reviewer: ```suggestion
You can also set app authentication and add SSO modules to your app using modules from the Mendix Marketplace. For more information, see the [App Authentication](#authentication) section below. 
```

> ## 2 Security in Studio Pro {#security-in-studio-pro}

Reviewer: Would this be better as something like
**App Security**
As it is about securing the app?
You set it up within Studio Pro, but you are not actually making Studio Pro any more secure.

Author: I named it **App Security Set Up in Studio Pro** for now. Or shall I shorten it to **App Security**?

### 2.1 Security Levels

> If you want full security, you need to explicitly give access to forms, entities, microflows, and workflows before someone can access them. By default, no one can access anything. To make it easier to create prototypes and demos there are security levels that require less security settings than are needed for a production system.

Reviewer: ```suggestion
If you want your app to be fully secure, you need to explicitly give access to forms, entities, microflows, and workflows before end-users can access them. When you turn production security on, the default is that no one can access anything. To make it easier to create prototypes and demos there are security levels that require less security than are needed for a production system.
```

See [App Security](/refguide/app-security/) for a description of the security levels.

### 2.2 App vs. Module Security

> At the level of an app some global settings can be specified: the security level, the administrator account and whether or not to allow anonymous access.

Reviewer: ```suggestion
At the level of an app some global settings can be specified: the security level, the administrator account, and whether or not to allow anonymous access.
```

## 1 Introduction

Security in Mendix has several aspects. One aspect is that your app is a project to work on in the backend, develop it and collaborate on it with your colleagues as a team. Yet you may want to control what each team member can change in the backend and this is managed in the Developer Portal. For more information, see [Building an App as a Team](#team) section below. 

> Another aspect is controlling what end-users can access in the running app. If you want different end-users to see different parts of your running application and you want to prevent unauthorized access, these things are managed from Studio Pro. Access to forms, data and microflows can be limited to authorized users. For more information, see [App Security Set Up in Studio Pro](#security-in-studio-pro) section below.

Reviewer: ```suggestion
Secondly, you will want to control who can access your app once it is running. You will want to prevent unauthorized access to your running application. You may also want different end-users to see different parts of the app , so access to individual pages, data, and microflows can be limited to specific end-users. This is managed from within Studio Pro. For more information, see [App Security Set Up in Studio Pro](#security-in-studio-pro) section below.
```

{{% alert color="info" %}}
Security in Mendix apps does not include scanning files that end-users upload or download from your application for viruses and malware. For more information, see the [Scanning Uploaded Files for Malicious Content](/howto/security/best-practices-security/#scanning-for-malicious-content) section in *How to Implement Best Practices for App Security*. 
{{% /alert %}}

> The last aspect is that you can also set app authentication and add SSO modules to your app using modules from the Mendix Marketplace. For more information, see the [App Authentication](#authentication) section below. 

Reviewer: This isn't really another aspect - it is just an extension of app security.

```suggestion
There are different ways in which you can set up app authentication, including local authentication and using a federated IdP to provide single sign-on (SSO). For more information, see the [App Authentication](#authentication) section below. 
```

See [Module Security](/refguide/module-security/).

### 2.3 User Roles vs. Module Roles {#user-role}

> An end-user in a Mendix application has one or more user roles. These roles can be assigned from within the client when creating or editing a user. User roles are at the level of an app and can be edited in [App Security](/refguide/app-security/).

Reviewer: ```suggestion
An end-user in a Mendix application has one or more user roles. These roles can be assigned from within the client when creating or editing an end-user. User roles are at the level of an app and can be edited in [App Security](/refguide/app-security/).
```

> A user role is a combination of module roles. A user that signs into the system gets the access rights of all of their user roles and indirectly to the module roles that are contained by those user roles.

Reviewer: ```suggestion
A user role is a combination of module roles. An end-user that signs into the system gets the access rights of all of their user roles and indirectly to the module roles that are contained within those user roles.
```

> For example, you have an app with two modules: System and ProjectManagement (PM). The PM module has three module roles: TeamMember, TeamLeader and Administrator. And let us say that in this case, we only need two user roles because we do not need the distinction between team leaders and administrators. You define those two user roles and assign module roles to them. The table below shows which module roles are contained within the user roles. Note that you always need at least the User role in System.

Reviewer: ```suggestion
Say, for example, you have an app with two modules: System and ProjectManagement (PM). The PM module has three module roles: TeamMember, TeamLeader, and Administrator. In this case, you do not need to make a distinction between team leaders and administrators so you only need two user roles. You define those two user roles and assign module roles to them. The table below shows which module roles are contained within the user roles. Note that you always need at least the User role in the System module.
```

> | User Role 'TeamMember' | User Role 'TeamLeader' |

Reviewer: Would it be clearer here to say
```suggestion
| Module Roles for User Role 'TeamMember' | Module Roles for User Role 'TeamLeader' |
```

| --- | --- |
| System.User | System.User |
| ProjectManagement.TeamMember | ProjectManagement.TeamLeader |
|   | ProjectManagement.Administrator |

### 2.4 Entity Access vs. Page Access {#entity-vs-page-access}

> Per entity you can specify who can read or write what members (attributes and associations) under what circumstances. Using XPath constraints you can express powerful security behavior (for example, "an employee can only see orders created by the department they are a part of").

Reviewer: Don't really like "Per entity" as a formula. How about
```suggestion
You can specify for each entity who can read or write which members (attributes and associations) under what circumstances. Using XPath constraints you can express powerful security behavior (for example, "an employee can only see orders created by the department they are a part of").
```

Reviewer: (Similarly for "per page" in the next paragraph!)

> A combination of entity access and a page access is necessary because entities can also be accessed from microflows and custom widgets. Furthermore, you can express more advanced security through entity access.
> 
> See [Entity Access](/refguide/module-security/#entity-access).

Reviewer: You could combine these if you wanted - I think having the link indication will encourage people to click it!
```suggestion
A combination of entity access and a page access is necessary because entities can also be accessed from microflows and custom widgets. Furthermore, you can express more advanced security through [entity access](/refguide/module-security/#entity-access).
```

## 3 Building an App as a Team {#team}

> If you would like to invite your colleagues to build the app, you can manage **Team** in Developer Portal. Only team members who are invited to your app can access your app (as a project to collaborate on). You can assign app roles to them, however, those are different from end-user roles. An end-user role defines what users can access in a running app while team member roles define what they can access and change in the app in the backend while developing this app. For more information, see [Team](/developerportal/general/team/) in the Developer Portal Guide.

Reviewer: I think this is worth turning around so we define what team roles are first, then why they are different from end-user roles.

```suggestion
If you want to invite your colleagues to build your app, you can manage the **Team** in the Developer Portal. Only team members who are invited to your app can access it (as a project to collaborate on). You can assign app roles to them. These team member roles define what they can access and change in the app in the backend while developing the app and are different from end-user roles which define what users can access in a running app. For more information, see [Team](/developerportal/general/team/) in the *Developer Portal Guide*.
```

> ## 4 App Authentication {#authentication}

Author: Added this from the Eval guide

Reviewer: Good idea - do we need to get someone from Collaboration or Engagement to review this as well?

Reviewer: Probably not worth it for now - perhaps see if we get any feedback once it is published.

> The Mendix platform provides you with three approaches to the authentication of users of your app. Depending on your preference you can include modules in your app from the Mendix Marketplace for each of these approaches.

Reviewer: ```suggestion
The Mendix platform provides you with three approaches to the authentication of end-users of your app. Depending on your preference you can include modules in your app from the Mendix Marketplace for each of these approaches.
```

> The first approach is to have "local" authentication in your app. With this approach, the users of your app are managed by your app. Users are authenticated against their password that is also stored within the app. In this approach the app is self-supporting and independent of any existing IAM infrastructure. The Mendix Marketplace also provides modules to also add multi-factor authentication to the local login process.

Reviewer: ```suggestion
The first approach is to have "local" authentication in your app. With this approach, the end-users of your app are managed by your app. End-users are authenticated against their password that is also stored within the app. In this approach the app is self-supporting and independent of any existing IAM (Identity and Access Management) infrastructure. The Mendix Marketplace also provides modules to also add multi-factor authentication to the local login process.
```

> A second approach is to have users authenticate via the Mendix Identity Provider through single sign on ([SSO](/appstore/modules/mendix-sso/)) using their Mendix credentials. This is the default option for developers who deploy their application to the Mendix Cloud.

Reviewer: I think we are trying to move customers away from this method. Perhaps you could get @Karuna-Mendix to discuss this with Jaap and work out how best to present this.

```suggestion
A second approach is to have end-users authenticate via the Mendix Identity Provider through single sign on ([SSO](/appstore/modules/mendix-sso/)) using their Mendix credentials. This is the default option for developers who deploy their application to the Mendix Cloud.
```

Author: @Karuna-Mendix , this comment was automatically marked as resolved, because I accepted the suggestion, but the question is still open. Is it correct that we are trying to move customers away from the second approach?

> The third approach is to integrate your app with a third-party Identity Provider (IDP) external to your app.

Reviewer: All the rest of the paragraphs in this section are related to this third approach (which should, perhaps be the second approach) - so perhaps we should make the sub-bullets in a bulleted list, or make headings for each of the three approaches to make it clearer?

And IDP should be IdP - I'll confirm with Karuna and put it into the terminology guide!

Reviewer 2: - I agree with @MarkvanMents . It would be nice to have three subsections for three approaches. 
- IDP should be IdP. I'll update terminology guide.
- multi-factor authentication (MFA)

Author: I've introduced sub-sections and changed the spelling. I will confirm with Karuna if we need to get rid of the second approach.

Customers with an on-premises Active Directory can use Microsoft’s Active Directory Federation Services (ADFS) to get SSO in their Mendix app. If you have this, you can include the platform-supported SAML SSO module in your app.

> Needless to say, the [SAML SSO](/appstore/modules/saml) module can be used with any IDP that supports the SAML 2.0 protocol. For example, this module module can be used with IDPs such as Okta, Microsoft Azure AD, Microsoft Active Directory Federation Service (ADFS), Shibboleth IdP’s, ForgeRock, and the Dutch eIDAS schemes DigiD and eHerkenning.

Reviewer: Should we put the SAML SSO link in the previous paragraph on the first mention, rather than here?

And I think that the example list is more something for the Evaluation Guide than here - we only need to specify "any IdP that supports the SAML 2.0 protocol" here, I think.

Author: I've incorporated both suggestions.

> Mendix also supports the OAuth protocol with the platform-supported [OIDC SSO](/appstore/modules/oidc/) module. OIDC/OAuth can be used in consumer facing applications to achieve single sign-on with identity providers such as Amazon Cognito, Google, Microsoft, LinkedIn, Twitter, or Facebook. It can also be used to authenticate to APIs, including APIs from Mendix ‘back-end’ apps which provide services to other apps. And, if your corporate IDP supports OIDC, you can use it to authenticate your employees and automatically grant them roles in your apps.

Reviewer: I think we can miss out the example list of identity providers here as well.

Reviewer 2: It is mentioned as 'SAML' in the Marketplace and in the Docs as well; not SAML SSO. I think we can provide the cross link as [SAML](/appstore/modules/saml).

Author: I've got rid of the example and changed "SAML SSO"  to "SAML".

> Mendix applications that are deployed to SAP BTP can have SSO with SAP’s IDP. Mendix provides a platform supported XSUAA connector to implement this federation in your Mendix app. 

Reviewer: ```suggestion
Mendix applications that are deployed to SAP BTP can have SSO with SAP's IDP. Mendix provides a platform supported [XSUAA connector](/appstore/modules/sap/sap-xsuaa-connector/) to implement this federation in your Mendix app. 
```
